Remove hash type encoding

HyunggyuJang opened the pull request:
Hash types were stored as hash type definition in the environment `Env.t`. 
But, there were only accessed by `Typetexp`. 

This PR moves their definitions to the corresponding class types where they are directly accessed.

We also generalized `Ctype.apply` and updated the comment.

Written with @garrigue and @t6s.

This PR completes #11568.

HyunggyuJang committed:
Remove hash type encoding

HyunggyuJang committed:
Bootstrap

HyunggyuJang committed:
Fix failing test due to id change

HyunggyuJang committed:
Update Changes

Update changes

garrigue commented on typing/types.mli:
```diff
@@ -565,6 +565,7 @@ type class_type_declaration =
   { clty_params: type_expr list;
     clty_type: class_type;
     clty_path: Path.t;
+    clty_ty: type_declaration;
```
rename to `clty_hash_type` (* object type with an open row *)

HyunggyuJang replied:
Reflected at https://github.com/ocaml/ocaml/pull/11569/commits/cebcd2698ffd3f4321ac9486a652fad95339cf1c

gasche commented on typing/types.mli:
```diff
@@ -565,6 +565,7 @@ type class_type_declaration =
   { clty_params: type_expr list;
     clty_type: class_type;
     clty_path: Path.t;
+    clty_ty: type_declaration;
```
Could we have a documentation comment that explains what this type declaration is?

Florian suggests "the type of the open row in this class type, used in `#t`" (if I understood correctly)

HyunggyuJang replied:
Will https://github.com/ocaml/ocaml/pull/11569/commits/cebcd2698ffd3f4321ac9486a652fad95339cf1c suffice?

gasche replied:
I believe so. Thanks!

gasche commented on typing/typeclass.ml:
```diff
@@ -1582,17 +1580,19 @@ let class_infos define_class kind
+        Ctype.newty(Tobject (tv, ref (Some (Pident ty_id, tv :: params)))) in
+      let ty1 = dummy_class_type cl_params
+      and ty2 = dummy_class_type cl_params' in
+      raise(Error(cl.pci_loc, env, Bad_parameters (ty_id, ty1, ty2)))
```
It is unclear that this error c ase can happen (@garrigue believes that it cannot). We proposed to use an `assert false` here -- and possibly in the similar error-handler below, and wait to see if someone ever encounters this. (Instead of having a not-really-meaningful error generated.)

gasche replied:
Actually @Octachron might have an example

```
# class ['a] c = object constraint 'a = int method m (x: bool c) = () end;;
Error: The abbreviation c is used with parameters bool c
       which are incompatible with constraints int c
# class ['a] c = object constraint 'a = int method m (x: bool #c) = () end;;
Error: The abbreviation #c is used with parameters bool #c
       which are incompatible with constraints int #c
```

Octachron replied:
I confirm that the second example reaches this case. And currently, there is a slight regression in the error message due to the use of the dummy class type:
```
Error: The abbreviation c is used with parameters bool #c/2
          which are incompatible with constraints int #c/2
```

gasche replied:
I propose to change the `Bad_parameters` error to take as argument the list of parameters (here two one-element lists: [`bool`] and [`int`]) instead of the application of the parameters to a type constructor.

Example outputs:

```
Error: The abbreviation #c is used with parameter(s) bool
          which are incompatible with constraint(s) int
```

In the multi-parameter case:


```
Error: The abbreviation #c is used with parameter(s) (bool, 'b)
          which are incompatible with constraint(s) (int, 'b)
```

HyunggyuJang replied:
Fixed as of https://github.com/ocaml/ocaml/pull/11569/commits/ea73b17eb8b84777ceaa8185f241dcc5ebf7c93b

HyunggyuJang replied:
Also added test case reflecting your suggestion, @gasche @Octachron

gasche replied:
Thanks! Could you add an example raising this error in the testsuite (for example the eexample of @Octachron?). This would let us remember how to reach this case, and show what output we get from your modified error-message printer.

HyunggyuJang replied:
Yep, already handled that at https://github.com/ocaml/ocaml/pull/11569/commits/071f1945c0a484816c8a1834fa3588dd810a69c2 (or https://github.com/ocaml/ocaml/commit/99372d64dc1c69eeb6b10743677aef42ca192789)

gasche replied:
Currently your testcase prints "The abbreviation c" instead of "The abbreviation #c". Is there a way to get `#c` instead?

HyunggyuJang replied:
Ah, right. I'll fix that.

HyunggyuJang replied:
We need to prefix hash (`#`) to either `ty_id` or when printing it by passing some boolean option for it.
I wonder which way would be better. Any suggestion?

HyunggyuJang replied:
Gone with more refined error case: https://github.com/ocaml/ocaml/pull/11569/commits/24805ce763d9832de3e3cc2f14f3ccfb1c2bc6cc

gasche replied:
Yes, I believe this is a correct approach.

garrigue commented on typing/typeclass.ml:
```diff
@@ -1582,17 +1580,19 @@ let class_infos define_class kind
-      let constr = Ctype.newconstr (Path.Pident cl_id) params in
+      let tv = Ctype.newvar () in
+      let constr =
+        Ctype.newty(Tobject (tv, ref (Some (Pident ty_id, tv :: params)))) in
```
And here also.

HyunggyuJang replied:
Should we amend this case also?

gasche replied:
This would be nice, but I don't see a way to modify the `Abbrev_type_clash` error that works well for the present use-case and also for the other use-cases for this error. Maybe you need to define a new error, different from `Abbrev_type_clash`, for this case, that does not require the first `type_expr` parameter (but takes a class name for example?). This could also let you produce a more precise/clear error message.

HyunggyuJang replied:
Like https://github.com/ocaml/ocaml/pull/11569/commits/b8cd3dda44289af0b1766c1ce0d4f895363db132? Sorry, I couldn't come up with better error naming and message for this case. Any thought? @garrigue

gasche replied:
Actually, thinking more about this, I think that my previous advice was wrong: to print a decent explanation for the error message, we really want to print a type looking like `(foo, bar) #c`, and then its expansion. In the case of `Bad_parameters` we had a natural formulation of the error explanation that would only mention `#c`, and not the whole type expression. But here we want the whole type expression, and the only way to do it is to do what you were previously doing in your patch, that is

```
      let constr =
        Ctype.newty(Tobject (tv, ref (Some (Pident ty_id, tv :: params)))) in
```

(question: why are you adding `tv` to the params?)

With the current state of my understanding, reverting to the previous approach that you had implemented seems best.

gasche replied:
My question is answered by types.mli:

```ocaml
  | Tobject of type_expr * (Path.t * type_expr list) option ref
  (** [Tobject (`f1:t1;...;fn: tn', `None')] ==> [< f1: t1; ...; fn: tn >]
      f1, fn are represented as a linked list of types using Tfield and Tnil
      constructors.

      [Tobject (_, `Some (`A.ct', [t1;...;tn]')] ==> [(t1, ..., tn) A.ct].
      where A.ct is the type of some class.

      There are also special cases for so-called "class-types", cf. [Typeclass]
      and [Ctype.set_object_name]:

        [Tobject (Tfield(_,_,...(Tfield(_,_,rv)...),
                         Some(`A.#ct`, [rv;t1;...;tn])]
             ==> [(t1, ..., tn) #A.ct]
        [Tobject (_, Some(`A.#ct`, [Tnil;t1;...;tn])] ==> [(t1, ..., tn) A.ct]

      where [rv] is the hidden row variable.
  *)
```

gasche replied:
Note: there is a nicer way to define this `constr` type expression:

```ocaml
let constr = Ctype.newobj (Ctype.newvar ()) in
Ctype.set_object_name ty_id params constr;
```

HyunggyuJang replied:
Thanks. Reflected at https://github.com/ocaml/ocaml/pull/11569/commits/f8f3ac49137999e38e8a0be5e99fc6600f95fa76

gasche commented on typing/typetexp.ml:
```diff
@@ -300,7 +291,7 @@ and transl_type_aux env policy styp =
         (List.combine stl args) params;
       let ty_args = List.map (fun ctyp -> ctyp.ctyp_type) args in
-      let ty = Ctype.expand_head env (newconstr path ty_args) in
+      let ty = Ctype.apply ~use_current_level:true env params body ty_args in
```
I reviewed the typetexp.ml part of this PR, with Jacques' explanations. It is tricky but I approve of this part of the change.

HyunggyuJang committed:
clty_ty -> clty_hash_type

HyunggyuJang committed:
Better error messages

Add test for error reporting

Pacify hygine

class_type specific abbrev error report

Update test

Reflect review

Better error message for bad parameters of class type

Revert "class_type specific abbrev error report"

This reverts commit b8cd3dda44289af0b1766c1ce0d4f895363db132.

Fix class type ident for Abbrev_type_clash

Add test for Abbrev_type_clash

Reflect review: Better error messaging

Reflect review: better error message for Bad_parameters

gasche commented on testsuite/tests/typing-objects/Tests.ml:
```diff
@@ -137,8 +137,8 @@ Lines 1-4, characters 0-3:
-Error: The abbreviation c is used with parameters bool c
-       which are incompatible with constraints int c
+Error: The abbreviation c is used with parameter(s) bool
+       which are incompatible with constraint(s) int
```
I would use "the class name c" instead of "the abbreviation c". Otherwise the error-message is fine, thanks!

HyunggyuJang replied:
Thanks! Reflected (actually pointed part comes from `Bad_parameters`)

Octachron replied:
`class name` ⇒ `class type`?

HyunggyuJang replied:
https://github.com/ocaml/ocaml/pull/11569/commits/597b4e6efd68290dcafe173ae46bba7390f4bc54 and https://github.com/ocaml/ocaml/pull/11569/commits/24805ce763d9832de3e3cc2f14f3ccfb1c2bc6cc should resolve it.

gasche commented on typing/typeclass.ml:
```diff
@@ -1449,16 +1445,16 @@ let temp_abbrev loc env id arity uid =
-  let (cl_params, cl_ty, env) = temp_abbrev cl.pci_loc env cl_id arity uid in
+  let (obj_params, obj_ty, obj_td) = temp_abbrev cl.pci_loc arity uid in
+  let env = Env.add_type ~check:true obj_id obj_td env in
+  let (cl_params, cl_ty, cl_td) = temp_abbrev cl.pci_loc arity uid in
```
I don't understand why you needed to change this code here to add `obj_td` to the environment, if it was not needed before. Can you explain? (Isn't there a risk that this "temporary abbreviation" remains (shadowed) in the final environment? We have been seeing `#c/1` and `#c/2` in some error messages and I wonder if it is related.)

HyunggyuJang replied:
Previously, `temp_abbrev` didn't return `cl_td`, which is needed for dummy `clty_hash_type`. Plus, do not need to treat `#t` type explicitly, we don't need to add dummy hash type to `env` anymore; so we decided not to modify `env` from `temp_abbrev`.

As a result, `obj_` related manipulations are intact (work as before), so it shouldn't be any issue.

gasche replied:
Ah, I missed this. It's not that you are adding one item to the env, it is that before you were adding one (inside the function) and now you are adding only one. Thanks for the clarification!

gasche reviewed (commented):
(During our meeting we said @Octachron would review this one but in fact I did a review.)

I'm very close to a full review of this PR. We discussed the error-cases quite a bit, we don't have the final version but this is converging. I have a last question on `initial_env`, see below.

Octachron commented on typing/typeclass.ml:
```diff
@@ -2046,13 +2026,21 @@ let report_error env ppf = function
-        "@[The abbreviation %a@ is used with parameters@ %a@ \
-           which are incompatible with constraints@ %a@]"
+        "@[The class type #%a@ is used with parameter(s)@ %a@ \
+           which are incompatible with constraint(s)@ %a@]"
```
I think it might be worth it to specify that it is the constraints on the class type parameter that are not satisfied:
```
 "@[The class type #%a@ is used with parameter(s)@ %a@ \
      which are incompatible with the constraint(s) on the class type parameters:@ %a@]"
```

or maybe even:

```
 "@[The class type #%a@ is used with parameter(s)@ %a,@ \
      whereas the class type definition@ constrains@ those parameters to be@ %a@]"
```

HyunggyuJang replied:
Thanks for the suggestion! Reflected!

gasche reviewed (approved):
I am now fully convinced that the PR is fine and can be merged. Approved.
Thanks for the nice (and not easy) work!

Remaining items:
- fill out the Changes entry
- decide whether you want to rebase/squash the commits on your end (otherwisee they will get squashed in just one commit)
- see what you want to do with the suggestions of @Octachron on the error message wording

HyunggyuJang commented:
@gasche I've cleaned up; will it work for you?

gasche commented:
Thanks! I merged the PR.

(The PR contains a bootstrap commit. I checked that redoing a bootstrap from that bootstrap commit gives the same result as from a local bootstrap on my machine.)
